### PR TITLE
Sets up Cloud Build to run our init actions tests on PRs

### PR DESCRIPTION
**Pull request by estebanlau**
Sets up Cloud Build infrastructure for triggering tests through GitHub App Triggers.

The tests triggered depending on the modified init actions.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (medb) [inline Dockerfile]: Remove extra spaces?

File: bigtable/test_bigtable.py [hunk 2]
```diff
@@ -108,5 +108,4 @@ def test_bigtable(self, configuration, dataproc_version, machine_suffixes):
 
 
 if __name__ == '__main__':
-    unittest.main()
```
Comment (medb) [inline bigtable/test_bigtable.py]: Revert this change?

File: cloudbuild.yaml [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+# Create docker image from regular Dockerfile
+- name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '--tag=gcr.io/google.com/hadoop-cloud-dev/init-actions-image', '.']
+# Run tests from the repo, based on the branch name
+- name: 'gcr.io/google.com/hadoop-cloud-dev/init-actions-image'
+  entrypoint: 'bash'
+  args:
+  - '-c'
+  - |
+    [[ "$BRANCH_NAME" == .*bigtable.* ]] && python bigtable/test_bigtable.py || exit 0
```
Comment (medb) [inline cloudbuild.yaml]: Is it possible to make this generic instead of copy/pasting for each test/init-action?

File: cloudbuild.yaml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  - |
+    [[ "$BRANCH_NAME" == .*tez.* ]] && python tez/test_tez.py || exit 0
+- name: 'gcr.io/google.com/hadoop-cloud-dev/init-actions-image'
+  entrypoint: 'bash'
+  args:
+  - '-c'
+  - |
+    [[ "$BRANCH_NAME" == .*tony.* ]] && python tony/test_tony.py || exit 0
+
+# images:
```
Comment (medb) [inline cloudbuild.yaml]: Is this comment necessary?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+COPY . /
+RUN ls
+
+RUN apt-get -y update
+RUN apt-get -y install python3-pip
+
+ENV PATH=$PATH:/builder/google-cloud-sdk/bin/
+ENV PYTHONPATH /
+
+RUN pip install -r requirements.txt
```
Comment (medb) [inline Dockerfile]: Rename `requirements.txt` to `requirements.pip`?

File: cloudbuild.yaml [hunk 6]
```diff
@@ -0,0 +1,111 @@
+steps:
+
+# Create docker image from regular Dockerfile
+- name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '--tag=gcr.io/google.com/hadoop-cloud-dev/init-actions-image', '.']
+# Run tests from the repo, based on the branch name
```
Comment (medb) [inline cloudbuild.yaml]: Instead of running all test lets run tests only for modified init actions. We may need to create shell or python script that does this and execute it here.

File: cloudbuild.yaml [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  args:
+  - '-c'
+  - |
+    [[ "$BRANCH_NAME" == .*solr.* ]] && python solr/test_solr.py || exit 0
+- name: 'gcr.io/google.com/hadoop-cloud-dev/init-actions-image'
+  entrypoint: 'bash'
+  args:
+  - '-c'
+  - |
+    [[ "$BRANCH_NAME" == .*starburst.* ]] && python starburst-presto/test_presto.py || exit 0
```
Comment (findepi) [inline cloudbuild.yaml]: Are you intentionally ignoring the exit code from python?

File: Dockerfile (hunk 1 above)
Comment (estebanlau) [reply Dockerfile]: Done.

File: bigtable/test_bigtable.py (hunk 2 above)
Comment (estebanlau) [reply bigtable/test_bigtable.py]: Done.

File: cloudbuild.yaml (hunk 4 above)
Comment (estebanlau) [reply cloudbuild.yaml]: Removed.

File: Dockerfile (hunk 5 above)
Comment (estebanlau) [reply Dockerfile]: Changed the extension.

File: cloudbuild.yaml (hunk 7 above)
Comment (estebanlau) [reply cloudbuild.yaml]: I was previously making sure that the steps were all executed in case one of the tests failed. I now changed the step to only fail if the python test fails. If the python test is simply not run if the branch name doesn't match, the step still succeeds.

Comment (estebanlau) [reply cloudbuild.yaml]: The 'waitFor' attribute defines the dependencies of each step. In this case, the tests run after the container is built.

File: cloudbuild.yaml (hunk 6 above)
Comment (estebanlau) [reply cloudbuild.yaml]: In here, only certain init actions are run (based on the branch name).

I propose changing this approach to having several defined set of build steps (different yaml files), each representing a given feature to be tested and triggered only when their respective files are modified. The downside is that we'll have many yaml files (one for each feature).

Comment (estebanlau) [reply cloudbuild.yaml]: This functionality is supported by GitHub App Triggers Alpha and we would end up with multiple triggers defined in GCP.

Comment (medb) [reply cloudbuild.yaml]: I think that this is nor reliable, because this convention is not obvious and requires human attention to enforce branch names.

What we need to do is to have script that diffs PR branch against master branch and after that runs tests for all init actions that modified by PR. Is it feasible to implement this with Cloud Build?

Comment (estebanlau) [reply cloudbuild.yaml]: Yes, I've been working on this approach. I can run a different set of steps based on files modified. CB GitHub App Triggers supports this by triggering builds depending on files included in a given PR. I'll have to define each trigger in GCP.

File: cloudbuild.yaml (hunk 3 above)
Comment (estebanlau) [reply cloudbuild.yaml]: Removed this file in place of multiple config files. Instead of deciding whether to run a given test based on the branch name as shown here, now we handle that logic at the trigger level in GCP.

File: cloudbuild.yaml (hunk 6 above)
Comment (medb) [reply cloudbuild.yaml]: Is CB GitHub App Triggers allow to configure this in generic way? i.e if `init-action-folder` is modified run `init-action-folder/test.py`? I'm asking, because I don't want to modify them each time we are adding new init action.

File: bigtable/bigtable_cloudbuild.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+steps:
+- name: 'gcr.io/cloud-builders/docker'
```
Comment (medb) [inline bigtable/bigtable_cloudbuild.yaml]: I don't think that this is a good approach to copy/paste these files in all init actions if instead we can create a single script that will execute relevant tests automatically.

File: cloudbuild.yaml (hunk 6 above)
Comment (estebanlau) [reply cloudbuild.yaml]: Yes, it allows us to define a generic path (e.g. when bigtable/** changes, we run bigtable/test_bigtable.py). If no changes are made to the files in the init action folder, its test is not triggered. For each init action, a generic rule like this will need to be defined.

File: bigtable/bigtable_cloudbuild.yaml (hunk 8 above)
Comment (estebanlau) [reply bigtable/bigtable_cloudbuild.yaml]: Are you suggesting using git diff to determine the modified files instead of using the built-in CB functionality that does this for us? If so, the challenges that come to mind are:

- Authenticating whoever is making the changes
- Determining how to git clone a version of the code that contains the changes we want to test. Git cloning from GoogleCloudPlatform/dataproc-initialization-actions would not contain our changes and depending on the user, the repo name would differ.

Comment (estebanlau) [reply bigtable/bigtable_cloudbuild.yaml]: I also thought that having every config copying the code seemed inconvenient, but I assumed that usually not many init action changes are included in a given PR (most commonly only 1?) so the building would only happen once for a given PR.

Comment (medb) [reply bigtable/bigtable_cloudbuild.yaml]: I think we talk past each other, let me describe what I mean with more details:

1. Create a single script that diffs PR changes and runs relevant tests based on them
2. Run this script as entry point of CB integration - I assume that this script will have access  to CB credentials to run tests and will run in git repo context inside CB docker container, so it will have access to git history to diff PR changes and determine what tests it needs to run.

Is this approach not feasible with CB?

File: cloudbuild.yaml (hunk 6 above)
Comment (medb) [reply cloudbuild.yaml]: I guess this is not generic enough comparing to what I thought, because it requires per-folder/init action pattern definition.

File: bigtable/bigtable_cloudbuild.yaml (hunk 8 above)
Comment (estebanlau) [reply bigtable/bigtable_cloudbuild.yaml]: It should be possible with CB. I'll look more into this to confirm and work on implementing this approach instead.

Comment (estebanlau) [reply bigtable/bigtable_cloudbuild.yaml]: CB now uses a single shell script that runs our tests as the entry point.

File: presubmit.sh [hunk 9]
```diff
@@ -1,92 +1,58 @@
...
-	return 1
-}
+declare -a DIRECTORIES_TO_TEST
+for dir in */
+do
+  # skip not init action changes
+  if [[ $dir =~ ^(integration_tests/|util/)$ ]]; then
+    continue
+  fi
+  if contains "$dir"; then
```
Comment (estebanlau) [inline presubmit.sh]: Nice shell script wizardry.

Does this account for the edge case where, for instance, starburst-presto is modified and presto is not? In this case, only starburst-presto/test_presto.py should be triggered.

Comment (medb) [reply presubmit.sh]: Good catch, modified patter to match prefix from the beginning of the string instead of contains.

**Comment by medb** [comment]
LGTM after @functicons comment addressed.

**Comment by estebanlau** [comment]
@functicons I'm working on moving the CB components within their own directory and including a README.md

**Comment by estebanlau** [comment]
I'm testing CB with the new script. My forked repo is currently integrated with CB for testing the triggers; once I verify that CB runs as expected with the new script, I'll need the credentials to the GoogleCloudPlatform/dataproc-initialization-actions repo to configure it in the google.com:cloud-dataproc-ci project or someone with the necessary credentials could do it.

File: cloud-build/README.MD [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# Continuous Integration via Cloud Build
+
+Cloud Build is configured in our GCP project (google.com:cloud-dataproc-ci) to trigger whenever there are PRs from a fork to the main repository. We are using [GitHub App Triggers Alpha](http://go/gcb-github#github-app-triggers) to integrate with GitHub.
```
Comment (functicons) [inline cloud-build/README.MD]: I don't think we want to expose the GCP project name publicly.

File: cloud-build/presubmit.sh [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  fi
+  mapfile -t tests_array < <(echo "${tests}")
+  ALL_TESTS+=("${tests_array[@]}")
+done
+echo "Tests: ${ALL_TESTS[*]}"
+
+# Run tests of the init actions that were modified
+python -m unittest "${ALL_TESTS[@]}"
+
+exit $?
```
Comment (functicons) [inline cloud-build/presubmit.sh]: Is this line necessary?

File: cloud-build/presubmit.sh [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    exit 1
+  fi
+  mapfile -t tests_array < <(echo "${tests}")
+  ALL_TESTS+=("${tests_array[@]}")
+done
+echo "Tests: ${ALL_TESTS[*]}"
+
+# Run tests of the init actions that were modified
+python -m unittest "${ALL_TESTS[@]}"
+
```
Comment (functicons) [inline cloud-build/presubmit.sh]: What if ALL_TESTS is empty?

File: Dockerfile [hunk 13]
```diff
@@ -0,0 +1,18 @@
+FROM gcr.io/cloud-builders/gcloud
```
Comment (functicons) [inline Dockerfile]: Outside the context of this PR, it is not clear what this Dockerfile is for, consider adding comments. Also, why this file is not under cloud-build?

Comment (estebanlau) [reply Dockerfile]: The docker build uses the Dockerfile found at the root of the repo. I tried to place it in cloud-build and use that moved Dockerfile instead by setting the 'dir' attribute of the step, but that made it impossible to access/copy the init actions from the parent directory.

Comment (estebanlau) [reply Dockerfile]: I'm looking more into this.

Comment (estebanlau) [reply Dockerfile]: The Dockerfile is inside now.

File: cloud-build/presubmit.sh (hunk 12 above)
Comment (estebanlau) [reply cloud-build/presubmit.sh]: If ALL_TESTS is empty, this means that no init action files were modified (so no tests will run) and the python command will return success.

Comment (medb) [reply cloud-build/presubmit.sh]: Actually, if `ALL_TESTS` is empty all tests will be run. I will add if condition.

File: cloud-build/cloudbuild.yaml [hunk 14]
```diff
@@ -0,0 +1,7 @@
+steps:
+# Create docker image from regular Dockerfile
+- name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '--tag=gcr.io/google.com/hadoop-cloud-dev/init-actions-image', '-f', 'cloud-build/Dockerfile', '.']
```
Comment (medb) [inline cloud-build/cloudbuild.yaml]: Is it possible to have the same path for `args` and `name` tags?

File: cloud-build/presubmit.sh (hunk 11 above)
Comment (medb) [reply cloud-build/presubmit.sh]: Removed

File: cloud-build/cloudbuild.yaml (hunk 14 above)
Comment (estebanlau) [reply cloud-build/cloudbuild.yaml]: The name refers to the base image that is being used and that has an entrypoint. In this case, gcr.io/cloud-builders/docker contains docker functionality.

File: cloud-build/README.MD (hunk 10 above)
Comment (estebanlau) [reply cloud-build/README.MD]: Removed.

File: cloud-build/presubmit.sh [hunk 15]
```diff
@@ -16,47 +16,45 @@ echo "Changed files: ${CHANGED_FILES[*]}"
...
 done
-echo "Tests: ${ALL_TESTS[*]}"
+echo "Tests: ${TESTS_TO_RUN[*]}"
 
 # Run tests of the init actions that were modified
-python -m unittest "${ALL_TESTS[@]}"
-
-exit $?
+if [[ $RUN_ALL_TESTS == true ]]; then
+  python -m unittest
```
Comment (estebanlau) [inline cloud-build/presubmit.sh]: Are you sure this runs all tests? Running this command in my machine results in no tests being run.

Comment (medb) [reply cloud-build/presubmit.sh]: Hm, I tested with `fastunit` and it runs all tests from the git repo root. Maybe there are difference between `unittest` and `fastunit`.

Anyway, we need to use `fastunit` because without it it will take hours to run all tests.

May you configure to use `fastunit` here?

File: cloud-build/cloudbuild.yaml (hunk 14 above)
Comment (medb) [reply cloud-build/cloudbuild.yaml]: Sorry for confusion, I meant `name` in a line below:
```
- name: 'gcr.io/google.com/cloud-dataproc-ci/init-actions-image'
```

Can we change `args` line to use `cloud-dataproc-ci` in the path:
```
  args: ['build', '--tag=gcr.io/google.com/cloud-dataproc-ci/init-actions-image', '-f', 'cloud-build/Dockerfile', '.']
```

File: cloud-build/presubmit.sh (hunk 15 above)
Comment (estebanlau) [reply cloud-build/presubmit.sh]: Yes, I will work on it.

File: cloud-build/cloudbuild.yaml [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+- name: 'gcr.io/google.com/cloud-dataproc-ci/init-actions-image'
```
Comment (functicons) [inline cloud-build/cloudbuild.yaml]: In the previous step, the project is hadoop-cloud-dev, but here it is cloud-dataproc-ci, why?

Comment (estebanlau) [reply cloud-build/cloudbuild.yaml]: Great catch! 

Wow, I spend hours yesterday debugging why the packages installed in the Dockerfile weren't accessible in the script. This seems to be it, thanks!

File: cloud-build/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# This Dockerfile spins up a container where presubmit tests are run.
+# Cloud Build orchestrates this process.
+
+FROM gcr.io/cloud-builders/gcloud
+
+# Copy everything into the container
+COPY . /
```
Comment (functicons) [inline cloud-build/Dockerfile]: I guess you want to copy the current project to a non / directory, e.g.: `COPY . /dataproc-init-actions/`

File: cloud-build/README.MD [hunk 18]
```diff
@@ -0,0 +1,12 @@
+# Continuous Integration via Cloud Build
+
+Cloud Build is configured in our GCP project (google.com:cloud-dataproc-ci) to trigger whenever there are PRs from a fork to the main repository. We are using GitHub App Triggers Alpha to integrate with GitHub.
```
Comment (functicons) [inline cloud-build/README.MD]: I think we should remove "google.com:cloud-dataproc-ci", that is sensitive information.

File: cloud-build/Dockerfile (hunk 17 above)
Comment (estebanlau) [reply cloud-build/Dockerfile]: Done.

File: cloud-build/cloudbuild.yaml [hunk 19]
```diff
@@ -0,0 +1,7 @@
+steps:
+# Create docker image from regular Dockerfile
+- name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '--tag=gcr.io/google.com/cloud-dataproc-ci/init-actions-image', '-f', 'cloud-build/Dockerfile', '.']
```
Comment (functicons) [inline cloud-build/cloudbuild.yaml]: Seem we can use variable $PROJECT_ID https://cloud.google.com/cloud-build/docs/configuring-builds/substitute-variable-values

Comment (estebanlau) [reply cloud-build/cloudbuild.yaml]: $PROJECT_ID would be substituted for "google.com:cloud-dataproc-ci" but we need "google.com/cloud-dataproc-ci" as part of our GCR path. See https://cloud.google.com/container-registry/docs/overview#domain-scoped_projects.

Comment (medb) [reply cloud-build/cloudbuild.yaml]: I have asked Andrey to create new project without `google.com:` prefix, maybe let's wait until it will be done.

File: fastunit/__init__.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
+"""
```
Comment (medb) [inline fastunit/__init__.py]: May we pull fastunit dependency from GitHub repo, instead of copying it in init actions repo? Maybe we need to create Docker image with pre-installed fastunit?

File: cloud-build/Dockerfile [hunk 21]
```diff
@@ -0,0 +1,20 @@
+# This Dockerfile spins up a container where presubmit tests are run.
+# Cloud Build orchestrates this process.
+
+FROM gcr.io/cloud-dataproc-ci/init-actions-image
```
Comment (medb) [inline cloud-build/Dockerfile]: How this image is created?

File: cloud-build/Dockerfile [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+
+# Copy everything into the container
+COPY . init-actions/
+WORKDIR init-actions/
+RUN ls -a
+
+RUN apt-get -y update && \
+    apt-get -y install python3-pip
+
+RUN python3.5 -m pip install --upgrade pip
```
Comment (medb) [inline cloud-build/Dockerfile]: Can we use just `python3` instead?

File: cloud-build/Dockerfile (hunk 21 above)
Comment (estebanlau) [reply cloud-build/Dockerfile]: I created it by manually running cloud build steps that pushed the image created from the Dockerfile.

File: cloud-build/Dockerfile (hunk 22 above)
Comment (estebanlau) [reply cloud-build/Dockerfile]: Yes.

File: cloud-build/Dockerfile (hunk 21 above)
Comment (medb) [reply cloud-build/Dockerfile]: Can we depend on default gcloud image instead? Or this image has some pre-installed software (fastunit, etc)? In general I think that we shouldn't depend on one-off images and instead put all boot strapping logic in this image.

Comment (estebanlau) [reply cloud-build/Dockerfile]: This image has fastunit pre-installed. I fixed the Dockerfile to base itself off the gcloud image and also included the fastunit directory so that it can be readily used.

Comment (estebanlau) [reply cloud-build/Dockerfile]: The downside is that the gcloud image doesn't have fastunit pre-installed so we have to copy the fastunit directory.

File: cloud-build/Dockerfile [hunk 23]
```diff
@@ -0,0 +1,20 @@
+# This Dockerfile spins up a container where presubmit tests are run.
+# Cloud Build orchestrates this process.
+
+FROM gcr.io/cloud-builders/gcloud
+
+# Copy everything into the container
+COPY . init-actions/
```
Comment (medb) [inline cloud-build/Dockerfile]: Cloud Build mounts GitHub repo to `/workspace` directory (current directory here). You need to copy it to new directory only if you are running multiple containers in the same build, otherwise you can just run presubmit in the workspace.

You can safely delete `COPY` and `WORKDIR` commands

File: fastunit/__init__.py (hunk 20 above)
Comment (medb) [inline fastunit/__init__.py]: We can just clone fastunit from GitHub in Docker container instead (without creating an image) by adding this command to `Docker` file and use it in pre-submit:
```bash
WORKDIR /fastunit
RUN git clone <fastunit repo> .
```

This way we will not need to create Docker image with pre-backed fastunit and will not need to include it into the init action repo.

File: cloud-build/Dockerfile (hunk 21 above)
Comment (medb) [reply cloud-build/Dockerfile]: I was thinking about cloning fastunit repo in Docker file as an alternative - added separate comment for this.

File: fastunit/__init__.py (hunk 20 above)
Comment (estebanlau) [reply fastunit/__init__.py]: Done.

File: integration_tests/setup.py [hunk 24]
```diff
@@ -0,0 +1,31 @@
+from setuptools import setup, find_packages
```
Comment (medb) [inline integration_tests/setup.py]: Seems like we don't need this file anymore.

File: cloudbuild/presubmit.sh [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+is_prefix() {
+  for file in "${CHANGED_FILES[@]}"; do
+    if [[ $file =~ ^$1 ]]; then
+      return 0
+    fi
+  done
+  return 1
+}
+
+determine_tests_to_run() {
```
Comment (functicons) [inline cloudbuild/presubmit.sh]: Can we add comments about how we detect changes? seems you cloned a repo, then somehow calculated the diff, but I don't understand.

Comment (medb) [reply cloudbuild/presubmit.sh]: Done.

File: oozie/test_oozie.py [hunk 26]
```diff
@@ -29,10 +29,10 @@ def __run_test_file(self, name):
             ("SINGLE", "1.1", ["m"]),
             ("SINGLE", "1.2", ["m"]),
             ("SINGLE", "1.3", ["m"]),
-            ("STANDARD", "1.1", ["m"]),
+            # ("STANDARD", "1.1", ["m"]),
```
Comment (estebanlau) [inline oozie/test_oozie.py]: Remove comments.

Comment (medb) [reply oozie/test_oozie.py]: Done.

File: cloudbuild/README.MD [hunk 27]
```diff
@@ -0,0 +1,15 @@
+# Continuous Integration via Cloud Build
```
Comment (functicons) [inline cloudbuild/README.MD]: Update this file with the K8S stuff.

Comment (medb) [reply cloudbuild/README.MD]: Done.

File: cloudbuild/README.MD [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+[Cloud Build](https://cloud.google.com/cloud-build/) is configured to trigger on
+`/gcbrun` comment from maintainer in PR to the `master` branch.
+
+Cloud Build triggers a set of build steps, configured in a `cloudbuild.yaml`
+config file. These steps are:
+
+1.  Create a presubmit Docker container with initialization actions code and
+    necessary dependencies installed; this includes git, gcloud, fastunit and
+    parametrized.
+1.  Push a presubmit Docker container to
```
Comment (functicons) [inline cloudbuild/README.MD]: s/a presubmit Docker container/the presubmit Docker container

File: cloudbuild/README.MD [hunk 29]
```diff
@@ -0,0 +1,21 @@
+# Continuous Integration via Cloud Build
```
Comment (functicons) [inline cloudbuild/README.MD]: Can you mention a bit about how we store/encode and use the SSH key?

File: cloudbuild/README.MD (hunk 28 above)
Comment (medb) [reply cloudbuild/README.MD]: Done.

File: cloudbuild/README.MD (hunk 29 above)
Comment (medb) [reply cloudbuild/README.MD]: Done.

File: cloudbuild/kubectl.sh [hunk 30]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (functicons) [inline cloudbuild/kubectl.sh]: nit: the filename kubectl.sh is a little bit confusing to me. I'd prefer something like run-presubmit-on-k8s.sh.

Comment (medb) [reply cloudbuild/kubectl.sh]: Done.
